nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py

```diff
@@ -0,0 +1,390 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+# Copyright 2019 The Google Research Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This file contains code artifacts adapted from the original implementation:
+https://github.com/google-research/google-research/blob/master/schema_guided_dst
+"""
+import collections
+import re
+from typing import List
+
+import numpy as np
+
+from nemo.collections.nlp.data.dialogue_state_tracking_generative.sgd.input_example import (
+    STATUS_ACTIVE,
+    SGDInputExample,
+)
+from nemo.core.classes import Dataset
+
+__all__ = ['DialogueSGDBERTDataset', 'DialogueBERTDataset']
+
+
+class DialogueBERTDataset(Dataset):
+    def __init__(self):
+        # The implementation for SGD is rather different from typical datasets
+        # Therefore, we leave this class unimplemented first
+        raise NotImplementedError()
+
+
+class DialogueSGDBERTDataset(Dataset):
+    '''
+    Dataset Class 
+        1. Performs Model-dependent (but Data-independent) operations (tokenization etc)
+        2. This can allow the same model preprocessing for multiple datasources
+        3. Users can configurate which labels to use for modelling 
+            (e.g. intent classification, slot filling or both together etc)
+    '''
+
+    def __init__(self, dataset_split: str, dialogues_processor: object, tokenizer, schemas, schema_config):
+        """ Constructor
+        Args:
+            dataset_split: dataset split`
+            dialogues_processor: Data generator for SGD dialogues
+        """
+        self.dataset_split = dataset_split
+        self.features = []
+        self.tokenizer = tokenizer
+        self.schemas = schemas
+        self.schema_config = schema_config
+        self.raw_features = dialogues_processor.get_dialog_examples(dataset_split)
+
+        for idx in range(len(self.raw_features)):
```

Comment:
have you tested how long this takes on SGD? since previously we used to cache the tokenized dataset, now we need to compute it from examples dict

Reply:
It takes around 7 minutes to process this. Since this takes some time, I have now implemented the caching function as part of the dataset class.